external-crates/move/crates/move-analyzer/src/symbols.rs

```diff
@@ -841,369 +886,732 @@ impl Symbols {
 
-        let modules = &typed_ast.unwrap().inner.modules;
+    // uwrap's are safe - this function returns earlier (during diagnostics processing)
+    // when failing to produce the ASTs
+    let typed_modules = &typed_ast.unwrap().inner.modules;
+
+    let mut mod_outer_defs = BTreeMap::new();
+    let mut mod_use_defs = BTreeMap::new();
+    let mut file_mods = BTreeMap::new();
+
+    for (pos, module_ident, module_def) in typed_modules {
+        let mod_ident_str = format!("{}", module_ident);
+        let (defs, symbols) = get_mod_outer_defs(
+            &pos,
+            &sp(pos, *module_ident),
+            module_def,
+            &files,
+            &file_id_mapping,
+            &file_id_to_lines,
+        );
 
-        let mut mod_outer_defs = BTreeMap::new();
-        let mut mod_use_defs = BTreeMap::new();
-        let mut file_mods = BTreeMap::new();
+        let cloned_defs = defs.clone();
+        let path = file_name_mapping.get(&cloned_defs.fhash.clone()).unwrap();
+        file_mods
+            .entry(
+                dunce::canonicalize(path.as_str()).unwrap_or_else(|_| PathBuf::from(path.as_str())),
+            )
+            .or_insert_with(BTreeSet::new)
+            .insert(cloned_defs);
 
-        for (pos, module_ident, module_def) in modules {
-            let (defs, symbols) = Self::get_mod_outer_defs(
-                &pos,
-                &sp(pos, *module_ident),
-                module_def,
-                &files,
-                &file_id_mapping,
-            );
+        mod_outer_defs.insert(mod_ident_str.clone(), defs);
+        mod_use_defs.insert(mod_ident_str, symbols);
+    }
 
-            let cloned_defs = defs.clone();
-            let path = file_name_mapping.get(&cloned_defs.fhash.clone()).unwrap();
-            file_mods
-                .entry(
-                    dunce::canonicalize(path.as_str())
-                        .unwrap_or_else(|_| PathBuf::from(path.as_str())),
-                )
-                .or_insert_with(BTreeSet::new)
-                .insert(cloned_defs);
+    eprintln!("get_symbols loaded file_mods length: {}", file_mods.len());
 
-            mod_outer_defs.insert(*module_ident, defs);
-            mod_use_defs.insert(*module_ident, symbols);
-        }
+    let mut references = BTreeMap::new();
+    let mut file_use_defs = BTreeMap::new();
 
-        eprintln!("get_symbols loaded file_mods length: {}", file_mods.len());
+    let mut parsing_symbolicator = ParsingSymbolicator {
+        mod_outer_defs: &mod_outer_defs,
+        files: &files,
+        file_id_mapping: &file_id_mapping,
+        file_id_to_lines: &file_id_to_lines,
+    };
 
-        let mut symbolicator = Symbolicator {
-            mod_outer_defs,
-            files,
-            file_id_mapping,
-            file_id_to_lines,
-            type_params: BTreeMap::new(),
-            current_mod: None,
+    parsing_symbolicator.prog_symbols(&parsed_ast.unwrap(), &mut references, &mut mod_use_defs);
```

Comment:
we appear to `unwrap` conditionally on `parsed_ast`, so wonder if it's useful to represent this as an `Option`, can we just use it directly without `Option` on parse success, am I missing something?

Reply:
(similar question for `typed_ast` actually, and maybe this is just the pattern of the existing code being followed)